File: python/src/aoc_2024_03.py
"""Advent of Code 2024 - Day 3."""

from re import findall
from sys import stdin

MUL_REGEX = r"mul\((\d+),(\d+)\)"
ALL_REGEX = r"(do\(\))|(don't\(\))|" + MUL_REGEX


def sum_uncorrupted(data: str) -> int:
    """Return sum of uncorrupted multiplications."""
    return sum(int(x) * int(y) for x, y in findall(MUL_REGEX, data))


def sum_enabled(data: str) -> int:
    """Return sum of enabled multiplications."""
    enabled = True
    total = 0
    for m in findall(ALL_REGEX, data):
        match m:
            case ("do()", "", "", ""):
                enabled = True
            case ("", "don't()", "", ""):
                enabled = False
            case ("", "", x, y):
                if enabled:
                    total += int(x) * int(y)
            case _:
                raise ValueError(f"Unexpected match {m}")
    return total


if __name__ == "__main__":
    data: str = stdin.read()
    print(sum_uncorrupted(data))
    print(sum_enabled(data))

File: rust/src/bin/aoc_2024_03.rs
// Advent of Code 2024 - Day 3.

use regex::Regex;
use std::io;

fn sum_uncorrupted(data: &str) -> i32 {
    Regex::new(r"mul\((\d+),(\d+)\)")
        .unwrap()
        .captures_iter(data)
        .map(|capture| capture[1].parse::<i32>().unwrap() * capture[2].parse::<i32>().unwrap())
        .sum()
}

fn sum_enabled(data: &str) -> i32 {
    let r = Regex::new(r"(do\(\))|(don't\(\))|mul\((\d+),(\d+)\)").unwrap();

    let mut enabled: bool = true;
    let mut total: i32 = 0;

    for capture in r.captures_iter(data) {
        if &capture[0] == "do()" {
            enabled = true;
        } else if &capture[0] == "don't()" {
            enabled = false;
        } else if enabled {
            total += capture[3].parse::<i32>().unwrap() * capture[4].parse::<i32>().unwrap();
        }
    }
    total
}

fn main() {
    let data: String = io::read_to_string(io::stdin()).unwrap();

    println!("{}", sum_uncorrupted(&data));
    println!("{}", sum_enabled(&data));
}
